fdbserver/BlobMigrator.actor.cpp

```diff
@@ -117,20 +143,61 @@ class BlobMigrator : public NonCopyable, public ReferenceCounted<BlobMigrator> {
-		wait(waitForDataMover(self));
-		state int oldMode = wait(setDDMode(self->db_, 0));
-		// Register as a storage server, so that DataDistributor could start data movement after
-		std::pair<Version, Tag> verAndTag = wait(addStorageServer(self->db_, self->interf_.ssi));
-		dprint("Started storage server interface {} {}\n", verAndTag.first, verAndTag.second.toString());
-
-		// Reassign key ranges to the storage server
-		// It'll restart DataDistributor so that internal data structures like ShardTracker, ShardsAffectedByTeamFailure
-		// could be re-initialized. Ideally it should be done within DataDistributor, then we don't need to
-		// restart DataDistributor
-		wait(unassignServerKeys(self, keys));
-		wait(assignKeysToServer(self, keys, self->interf_.ssi.id()));
-		wait(success(setDDMode(self->db_, oldMode)));
-		return Void();
+		state Future<ErrorOr<PrepareBlobRestoreReply>> replyFuture = Never();
+		state Future<Void> dbInfoChange = Void();
+		state Future<Void> delayTime = Void();
+		state int retries = 0;
+		state UID requestId;
+		loop {
+			choose {
+				when(wait(dbInfoChange)) {
+					if (self->dbInfo->get().distributor.present()) {
+						requestId = deterministicRandom()->randomUniqueID();
+						replyFuture = errorOr(self->dbInfo->get().distributor.get().prepareBlobRestoreReq.getReply(
+						    PrepareBlobRestoreRequest(requestId, self->interf_.ssi, keys)));
+						dbInfoChange = Never();
+						TraceEvent("BlobRestorePrepare", self->interf_.id())
+						    .detail("State", "SendReq")
+						    .detail("ReqId", requestId);
+					} else {
+						replyFuture = Never();
+						dbInfoChange = self->dbInfo->onChange();
+						TraceEvent(SevWarn, "BlobRestorePrepare", self->interf_.id()).detail("State", "WaitDD");
+					}
+				}
+				when(ErrorOr<PrepareBlobRestoreReply> reply = wait(replyFuture)) {
+					if (reply.isError()) {
+						TraceEvent("BlobRestorePrepare", self->interf_.id())
+						    .error(reply.getError())
+						    .detail("State", "Error")
+						    .detail("ReqId", requestId);
+
+						if (reply.getError().code() == error_code_restore_error) {
+							throw restore_error();
+						}
+					} else if (reply.get().res == PrepareBlobRestoreReply::SUCCESS) {
+						TraceEvent("BlobRestorePrepare", self->interf_.id())
+						    .detail("State", "Success")
+						    .detail("ReqId", requestId);
+						return Void();
+					} else {
+						TraceEvent("BlobRestorePrepare", self->interf_.id())
+						    .detail("State", "Failed")
+						    .detail("ReqId", requestId)
+						    .detail("Code", reply.get().res)
+						    .detail("Retries", retries);
+					}
+
+					if (++retries > SERVER_KNOBS->BLOB_MIGRATOR_ERROR_RETRIES) {
+						throw restore_error();
+					}
+					delayTime = delayJittered(10.0);
+					dbInfoChange = Void();
+				}
+			}
+
+			wait(delayTime);
+		}
 	}
 
 	// Wait until all pending data moving is done before doing full restore.
```

Comment:
we don't need this function any more,  right?

Reply:
yeah, I should remove `waitForDataMover`